app/services/cache_service.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class CacheService
+  def self.cache(cache_key, options = {})
+    Rails.cache.fetch cache_key.to_s, options do
+      yield
+    end
+  end
+
+  # Yields a cached query, expired by the most recently updated record for a given class.
+  # E.g: if *any* Spree::Taxon record is updated, all keys based on Spree::Taxon will auto-expire.
+  def self.cached_data_by_class(cache_key, cached_class)
+    Rails.cache.fetch "#{cache_key}-#{cached_class}-#{latest_timestamp_by_class(cached_class)}" do
+      yield
+    end
+  end
```

Comment:
Rails's normal built-in cache magic works with single records. Under the hood it creates a cache key derived from the record's `:updated_at` value to auto-expire related caches when the record is updated.

What we're gaining here with `#cached_data_by_class` and `#latest_timestamp_by_class` is usable cache keys that auto-expire when _any_ records for a given class are updated. So for example we can cache some AMS that serializes _all taxons_, and it'll be invalidated when _any_ taxon is updated. So if nobody has updated a `Spree::Taxon` in the last 4 hours, and we've fragment-cached a big chunk of taxons AMS in a view (after it's been fetched, serialized and _rendered_) using `latest_timestamp_by_class(Spree::Taxon)` in the key, we can keep serving the same cached data the whole time instead of rebuilding it on every page load. :tada: